### PR TITLE
Make the notary version detectable

### PR DESCRIPTION
**Pull request by LaurentGoderre**
Syft is able to use the `ldflags` from go to properly detect the main version:

https://github.com/anchore/syft/blob/main/syft/pkg/cataloger/golang/parse_go_binary_test.go#L1034-L1039

In order to make this work, we would need to enable that feature in the Scout Scanner

### CODE REVIEW

**Review by tianon-sso — APPROVED** [review]
We'll also probably need to bump the Alpine version 😭

File: Dockerfile.template [hunk 1]
```diff
@@ -1,3 +1,31 @@
+FROM golang:1.19-alpine{{ .alpine }} AS build
+
+RUN apk add --no-cache git make
+
+ENV NOTARYPKG github.com/theupdateframework/notary
```
Comment (whalelines) [inline Dockerfile.template:5]:
```suggestion
ENV NOTARYPKG github.com/notaryproject/notary
```

File: Dockerfile.template [hunk 2]
```diff
@@ -1,3 +1,31 @@
...
+	echo "${TAG//v/}" > NOTARY_VERSION; \
+# https://github.com/notaryproject/notary/pull/1635
+	git fetch --depth 2 origin efc35b02698644af16f6049c7b585697352451b8; \
+	git -c user.name=foo -c user.email=foo@example.com cherry-pick -x efc35b02698644af16f6049c7b585697352451b8; \
+# https://github.com/notaryproject/notary/issues/1602 (rough cherry-pick of ca095023296d2d710ad9c6dec019397d46bf8576)
+	go get github.com/dvsekhvalnov/jose2go@v0.0.0-20200901110807-248326c1351b; \
+	go mod vendor; \
+# TODO remove for the next release of Notary (which should include efc35b02698644af16f6049c7b585697352451b8 & ca095023296d2d710ad9c6dec019397d46bf8576)
+# Make the version detectable by scanners
+	sed -i -r -E 's|(version.NotaryVersion=\$\(NOTARY_VERSION\))|\1 -X $(NOTARY_PKG)/version.Version=$(NOTARY_VERSION)|' Makefile; \
```
Comment (whalelines) [inline Dockerfile.template:23]:
```suggestion
	sed -i -E 's,(version\.NotaryVersion=\$\(NOTARY_VERSION\)),\1 -X $(NOTARY_PKG)/version.Version=$(NOTARY_VERSION),' Makefile; \
```

`-r` and `-E` are redundant, `-E` is more common.

I'd avoid using a regular expression special character, `|` in this case, as the substitution delimiter.

**Review by whalelines — COMMENTED** [review]
Can the rationale for moving from an intermediate image to a two-stage build be given in the PR description?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -10,10 +10,6 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
-      - run: |
-          docker build notary-builder --tag notary:builder
-          tag="$(docker run --rm notary:builder sh -c 'echo $TAG' | awk '{gsub(/^v/, ""); print}')"
-          docker tag notary:builder "notary:${tag}-builder"
```
[the remark this answers is not in the document]
Comment (LaurentGoderre) [reply .github/workflows/ci.yml:16]: https://github.com/docker-library/official-images/pull/15568#issuecomment-1843637747

File: Dockerfile.template [hunk 4]
```diff
@@ -1,3 +1,31 @@
...
+
+ENV NOTARYPKG github.com/theupdateframework/notary
+ENV TAG v{{ .version }}
+
+ENV GOFLAGS -mod=vendor
+
+WORKDIR /go/src/$NOTARYPKG
+RUN set -eux; \
+	git clone -b "$TAG" --depth 1 "https://$NOTARYPKG.git" .; \
+# In case the version in file doens't match the tag (like in 0.7.0)
```
[the remark this answers is not in the document]
Comment (LaurentGoderre) [reply Dockerfile.template:13]: When version 0.7.0 was release, the main file used in the code to determine the version wasn't bumped. 

This is the 0.7.0 commit where the version is wrong.

https://github.com/notaryproject/notary/blob/b0b6bfdd4933081e8d5ae026b24e8337311dd598/NOTARY_VERSION

File: Dockerfile.template (hunk 1 above)
Comment (LaurentGoderre) [reply Dockerfile.template:5]: This must match what is in the go manifest which in version 0.7.0 is still that one as per:

https://github.com/notaryproject/notary/blob/b0b6bfdd4933081e8d5ae026b24e8337311dd598/go.mod#L1

File: Dockerfile.template (hunk 2 above)
Comment (LaurentGoderre) [reply Dockerfile.template:23]: Why not the special character? I've seen it done like this many time

File: Dockerfile.template (hunk 1 above)
Comment (whalelines) [reply Dockerfile.template:5]: They are the same repo, theupdateframework/notary redirects to notaryproject/notary, as your link shows.

File: Dockerfile.template (hunk 2 above)
Comment (whalelines) [reply Dockerfile.template:23]: How would you use the special character? How would you escape it?

File: Dockerfile.template (hunk 1 above)
Comment (tianon-sso) [reply Dockerfile.template:5]: Right, but this is used in the Go modules, so it has to match what _Go_ thinks the module name is 🙃

File: Dockerfile.template (hunk 2 above)
Comment (tianon-sso) [reply Dockerfile.template:23]: I usually use `@` or `!` as my "non-`/` delimiter", but I don't think the exact choice matters that much here, right?  If we want to use `|` in our regex in the future, it's pretty easy to swap then. 🤷

File: Dockerfile.template (hunk 1 above)
Comment (LaurentGoderre) [reply Dockerfile.template:5]: Exactly! It's confusing I know

**Comment by LaurentGoderre** [comment]
We can undo those change when this lands: https://github.com/notaryproject/notary/pull/1704

File: Dockerfile.template (hunk 1 above)
Comment (thaJeztah) [reply Dockerfile.template:5]: Careful; the module wasn't renamed yet in the current release though (still uses `github.com/theupdateframework/notary`); https://github.com/notaryproject/notary/blob/v0.7.0/go.mod#L1

i.e. `github.com/notaryproject/notary` did not yet do a release

**Review by jonnystoten — APPROVED** [review]
:heart:

**Pull request merged**